x-pack/plugins/cases/server/connectors/cases/cases_connector.ts

```diff
@@ -38,5 +97,300 @@ export class CasesConnector extends SubActionConnector<
+
+    for (const oracleRecord of oracleRecords) {
+      const { alerts, grouping } = groupedAlertsWithOracleKey.get(oracleRecord.id) ?? {
+        alerts: [],
+        grouping: {},
+      };
+
+      const caseId = this.casesService.getCaseId({
+        ruleId: rule.id,
+        grouping,
+        owner,
+        spaceId,
+        counter: oracleRecord.counter,
+      });
+
+      casesMap.set(caseId, { caseId, alerts, grouping, oracleKey: oracleRecord.id });
+    }
+
+    return casesMap;
+  }
+
+  private async bulkGetOrCreateCases(
+    params: CasesConnectorRunParams,
+    casesClient: CasesClient,
+    groupedAlertsWithCaseId: Map<string, GroupedAlertsWithCaseId>
+  ): Promise<Map<string, GroupedAlertsWithCases>> {
+    const bulkCreateReq: BulkCreateCasesRequest['cases'] = [];
+    const casesMap = new Map<string, GroupedAlertsWithCases>();
+
+    const ids = Array.from(groupedAlertsWithCaseId.values()).map(({ caseId }) => caseId);
+    const { cases, errors } = await casesClient.cases.bulkGet({ ids });
+
+    for (const theCase of cases) {
+      if (groupedAlertsWithCaseId.has(theCase.id)) {
+        const data = groupedAlertsWithCaseId.get(theCase.id) as GroupedAlertsWithCaseId;
+        casesMap.set(theCase.id, { ...data, theCase });
+      }
+    }
+
+    if (errors.length === 0) {
+      return casesMap;
+    }
+
+    /**
+     * TODO: Throw/retry for other errors
+     */
+    const nonFoundErrors = errors.filter((error) => error.status === 404);
+
+    if (nonFoundErrors.length === 0) {
+      return casesMap;
+    }
+
+    for (const error of nonFoundErrors) {
+      if (groupedAlertsWithCaseId.has(error.caseId)) {
+        const data = groupedAlertsWithCaseId.get(error.caseId) as GroupedAlertsWithCaseId;
+
+        bulkCreateReq.push(this.getCreateCaseRequest(params, data));
+      }
+    }
+
+    /**
+     * TODO: bulkCreate throws an error. Retry on errors.
+     */
+    const bulkCreateCasesResponse = await casesClient.cases.bulkCreate({ cases: bulkCreateReq });
+
+    for (const res of bulkCreateCasesResponse.cases) {
+      if (groupedAlertsWithCaseId.has(res.id)) {
+        const data = groupedAlertsWithCaseId.get(res.id) as GroupedAlertsWithCaseId;
+        casesMap.set(res.id, { ...data, theCase: res });
+      }
+    }
+
+    return casesMap;
+  }
+
+  private getCreateCaseRequest(
+    params: CasesConnectorRunParams,
+    groupingData: GroupedAlertsWithCaseId
+  ) {
+    const { grouping } = groupingData;
+
+    const ruleName = params.rule.ruleUrl
+      ? `[${params.rule.name}](${params.rule.ruleUrl})`
+      : params.rule.name;
+
+    const groupingDescription = this.getGroupingDescription(grouping);
+
+    const description = `This case is auto-created by ${ruleName}. \n\n Grouping: ${groupingDescription}`;
+
+    const tags = Array.isArray(params.rule.tags) ? params.rule.tags : [];
+
+    /**
+     * TODO: Add grouping info to
+     */
+    return {
```

Comment:
what about the required custom fields?

Reply:
Great point. I need to think about it. We need the `default value` feature.